Add Cesium ion geocoder to CesiumIonClient::Connection

azrogers opened the pull request:
As described in #974, we want to make it easier for users to find the coordinates of locations of interest in our integrations, as an alternative to looking up the coordinates through some other service and entering them manually. The first step of this is completing #70, an implementation of the Cesium ion geocoder API, which is what this PR is for. It adds a `geocode` method to `CesiumIonClient::Connection` modeled on the CesiumJS geocoder API. This isn't a complete geocoder implementation supporting multiple services, but it will allow us to implement CesiumGS/cesium-unreal#99 to improve Cesium for Unreal's ease of use.

kring commented on CesiumIonClient/include/CesiumIonClient/Geocoder.h:
```diff
@@ -0,0 +1,131 @@
+      destination;
+
+  GeocoderFeature(
+      std::string& displayName_,
+      CesiumGeospatial::GlobeRectangle& destination_)
```
```suggestion
      const std::string& displayName_,
      const CesiumGeospatial::GlobeRectangle& destination_)
```

kring commented on CesiumIonClient/include/CesiumIonClient/Geocoder.h:
```diff
@@ -0,0 +1,131 @@
+      CesiumGeospatial::GlobeRectangle& destination_)
+      : displayName(displayName_), destination(destination_) {}
+  GeocoderFeature(
+      std::string& displayName_,
+      CesiumGeospatial::Cartographic& destination_)
```
```suggestion
      const std::string& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
```

kring replied:
Alternatively, you could do this:

```cpp
GeocoderFeature(
      std::string displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(std::move(displayName_)), destination(destination_) {}
```

Which would be slightly more efficient because you can optionally move the string into the constructor (moving a GlobeRectangle isn't any different from copying it).

Slightly more efficient still would be two constructors, but that's almost certainly not worth the noise:

```cpp
GeocoderFeature(
      std::string&& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(std::move(displayName_)), destination(destination_) {}

GeocoderFeature(
      const std::string& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(displayName_), destination(destination_) {}
```

kring replied:
Another possibility here is to just get rid of the constructors entirely for these simple value-holding types. Users could still create instances with almost the exact same syntax, except they'd have to use curly braces.

kring commented on CesiumIonClient/include/CesiumIonClient/Geocoder.h:
```diff
@@ -0,0 +1,131 @@
+   */
+  bool showOnScreen;
+
+  GeocoderAttribution(std::string& html_, bool showOnScreen_)
```
```suggestion
  GeocoderAttribution(const std::string& html_, bool showOnScreen_)
```

kring commented on CesiumIonClient/include/CesiumIonClient/Connection.h:
```diff
@@ -291,6 +292,21 @@ class CESIUMASYNC_API Connection {
+   * @param query The query to make.
+   */
+  CesiumAsync::Future<Response<GeocoderResult>> geocode(
+      const GeocoderProviderType& provider,
+      const GeocoderRequestType& type,
```
Generally we pass enums by value instead of const reference, cause they're just integers.

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -1,10 +1,15 @@
 #include "CesiumIonClient/Connection.h"
 
+#include "CesiumGeospatial/BoundingRegion.h"
+#include "CesiumGeospatial/Cartographic.h"
+#include "CesiumGeospatial/GlobeRectangle.h"
+#include "CesiumIonClient/Geocoder.h"
```
This file is inconsistent to begin with, but we should move toward `#include <CesiumHeader/Whatever.h>` instead of `#include "CesiumHeader/Whatever.h"`.

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -582,6 +588,85 @@ Defaults defaultsFromJson(const rapidjson::Document& json) {
+  auto featuresMemberIt = json.FindMember("features");
+  if (featuresMemberIt != json.MemberEnd() &&
+      featuresMemberIt->value.IsArray()) {
+    auto featuresIt = featuresMemberIt->value.GetArray();
```
This isn't an iterator, is it?

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -582,6 +588,85 @@ Defaults defaultsFromJson(const rapidjson::Document& json) {
+      featuresMemberIt->value.IsArray()) {
+    auto featuresIt = featuresMemberIt->value.GetArray();
+    for (auto& feature : featuresIt) {
+      const rapidjson::Value* pLabel =
+          rapidjson::Pointer("/properties/label").Get(feature);
```
Not a huge deal, but it'd be nice to move the `rapidjson::Pointer` construction outside the loop. No reason to do it repeatedly.

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -582,6 +588,85 @@ Defaults defaultsFromJson(const rapidjson::Document& json) {
+        continue;
+      }
+
+      std::string label(pLabel->GetString());
```
What if `properties.label` isn't a string? It'd be safer to either check `IsString`, or use `JsonHelpers::getStringOrDefault` (which will do this for you).

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -582,6 +588,85 @@ Defaults defaultsFromJson(const rapidjson::Document& json) {
+      const rapidjson::Value* pLabel =
+          rapidjson::Pointer("/properties/label").Get(feature);
+      if (!pLabel) {
+        SPDLOG_WARN("Missing label for geocoder feature");
```
We should generally avoid logging this sort of thing directly, because it's not accessible programmatically and in some cases it might be viewed as spam. The pattern used elsewhere in this file (use the `OrDefault` methods to silently fill in a default if the JSON doesn't match what we expect) isn't perfect, but better to stick with that for consistency.

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -582,6 +588,85 @@ Defaults defaultsFromJson(const rapidjson::Document& json) {
+    result.attributions.reserve(valueJson.Size());
+
+    for (rapidjson::SizeType i = 0; i < valueJson.Size(); ++i) {
+      const auto& element = valueJson[i];
```
It's ok to use `auto` for iterators, but this is more clear without:
```suggestion
      const rapidjson::Value& element = valueJson[i];
```

kring commented on CesiumIonClient/src/Connection.cpp:
```diff
@@ -1179,3 +1264,64 @@ Connection::tokens(const std::string& url) const {
+  requestUrl = CesiumUtility::Uri::addQuery(requestUrl, "text", query);
+
+  // Add provider type to url
+  switch (provider) {
+  case GeocoderProviderType::Bing:
+    requestUrl = CesiumUtility::Uri::addQuery(requestUrl, "geocoder", "BING");
+    break;
+  case GeocoderProviderType::Google:
+    requestUrl = CesiumUtility::Uri::addQuery(requestUrl, "geocoder", "GOOGLE");
+    break;
+  case GeocoderProviderType::Default:
+    requestUrl =
+        CesiumUtility::Uri::addQuery(requestUrl, "geocoder", "DEFAULT");
+    break;
+  }
```
CesiumJS seems to use lowercase names, and leaves the parameter out entirely for `Default`:
https://github.com/CesiumGS/cesium/pull/12299/files#diff-1a5903d86aa48fe1b27e5f0d8e5da57e1afc40ee0e1ee6029fded82c23f5b1aaR28

kring reviewed (commented):
Looks good @azrogers! Just some nitpicky comments here.

azrogers commented:
Request to get this finished from the forums: https://community.cesium.com/t/geocoder-for-unity/37435

azrogers commented:
@kring Addressed review comments!

azrogers commented:
Removed the constructors for `GeocoderFeature` and `GeocoderAttribution`, as you're right, they can just be constructed by braces - but this causes an issue on clang with `emplace_back`, which surfaced a hazy memory that that's the reason I added the constructors in the first place. Do you know a way around this error? @kring

kring commented:
> but this causes an issue on clang with emplace_back

Oh right, I don't think you can use emplace_back to call a brace-initializer "constructor". But if you construct it directly and then let the emplace_back move it into the vector, that works fine. I pushed up a change to do that.

kring commented on CesiumIonClient/test/TestConnection.cpp:
```diff
@@ -1,3 +1,6 @@
+#include "CesiumGeospatial/Cartographic.h"
```
Might as well fix the include style here.

kring reviewed (commented):
Thanks @azrogers, this looks great! Please just merge in main and then this is good to go. One super trivial comment below.

azrogers commented:
Made a CMake change to use `set(CESIUM_CLANG_TIDY_USE_THREADS ... CACHE ...)` instead of cmake_dependent_option, as apparently `option`/`cmake_dependent_option` only accept bool values, and the result was that clang-tidy was erroring when not specifying `CESIUM_CLANG_TIDY_USE_THREADS` on the command line as it was getting the flag `-jOFF`.

kring commented:
Thanks @azrogers!
